dlib/dnn/trainer.h

```diff
@@ -1020,6 +1027,23 @@ namespace dlib
                     sync_file_reloaded = true;
                     if (verbose)
                         std::cout << "Loss has been increasing, reloading saved state from " << newest_syncfile() << std::endl;
+
+                    // Are we repeatedly hitting our head against the wall? If so, then we
+                    // might be better off giving up at this learning rate, and trying a
+                    // lower one instead.
+                    if (prob_loss_increasing_thresh >= prob_loss_increasing_thresh_max_value)
+                    {
+                        std::cout << "(and while at it, also shrinking the learning rate)" << std::endl;
+                        learning_rate = learning_rate_shrink * learning_rate;
+                        steps_without_progress = 0;
+                        test_steps_without_progress = 0;
+                        // Empty out some of the previous loss values so that steps_without_progress and test_steps_without_progress
+                        // will decrease below iter_without_progress_thresh.  
+                        for (unsigned long cnt = 0; cnt < previous_loss_values_dump_amount+iter_without_progress_thresh/10 && previous_loss_values.size() > 0; ++cnt)
+                            previous_loss_values.pop_front();
+                        for (unsigned long cnt = 0; cnt < test_previous_loss_values_dump_amount+test_iter_without_progress_thresh/10 && test_previous_loss_values.size() > 0; ++cnt)
+                            test_previous_loss_values.pop_front();
+                    }
```

Comment:
This same bit of code it duplicated elsewhere.  Can you put this into a function (or actually two functions I think, one for the test and another for the normal loss values seems sensible) and call it instead of copying and pasting the code?

Reply:
Sure; fixed. (Yeah, this is my preference also.)